src/Starcounter.Database.Extensions/NestedTransactor.cs

```diff
@@ -229,12 +241,25 @@ protected override void LeaveContext(IDatabaseContext db, bool exceptionThrown)
 
         protected void CaptureException(Exception ex) => _current.Value.InnerException = ex;
 
-        void FailOuterIfInnerFailed()
+        private void FailOuterIfInnerFailed()
         {
-            if (_current.Value.InnerException != null)
+            if (_current.Value?.InnerException != null)
```

Comment:
Is it possible to have null `_current.Value` here?

Reply:
Yes, when leaving context of a read-only transaction.

Reply:
Handling read-only transactions in this decorator is questionable. The reason is that it uses Transaction.Current which we hoped we wouldn't use in public code as it contradicts to DI approach. I hoped we could implement rebasing decorator in palindrom specific code base. Just like TransactAndRebase. And make Palma code create a proper decorator chain.

Reply:
If I understand you correctly, your idea is that:

- Starcounter.Palindrom defines a decorator, which would restart parent read-only transaction, once a nested read-write transaction committed.
- PALMA defines a decorator, which would allow "nested" transactions run within the outer scope.

In this case, the PALMA decorator still needs to know if the current transaction is read-only or read-write, because running a "nested" transaction within a parent read-only transaction does not provide what PALMA needs.

Reply:
My idea is that Palma creates a decorator chain: NestedTransactor -> PalindromTransactor -> Built-in Transactor. Where PalindromTransactor is the one that rebases outer read-only transaction and relies on Transaction.Current.

Reply:
Then I understood your idea correctly, the thing is that `NestedTransactor` has to behave according to the parent transaction's state:

- Within a read-write transaction: execute provided delegate straight away.
- Within a read-only transaction: execute provided delegate with a new read-write transaction.

Do I miss something?

Reply:
>Then I understood your idea correctly, the thing is that NestedTransactor has to behave according to the parent transaction's state:

It's a hard question and if you consider this, it should match parent's transaction state with flags passed to `Transact` call.
But it's not specific to PALMA and I think PALAM will be fine with current NestedTransactor as Palindrom creates outer transaction for Plama with Transaction.Create and thus, NestedTransaction won't detect it and start an inner transaction whatsoever.

Reply:
In this case we will rely on the fact that Starcounter.Palindrom does not use `ITransactor` to create transactions. Not sure if it's a good or bad way, since we already rely on `Transaction.Current` in there, but sounds like a hack on top of a hack.

Reply:
>In this case we will rely on the fact that Starcounter.Palindrom does not use ITransactor to create transactions


If Palidrom ever starts using decorators, it shouldn't be using `NestedTransactor`. At least not in its current shape. Then we can consider fixing `NestedTransactor`. But the `PalidromTransactor` (which performs rebase) would be still the same. So I guess we can start with it and it will suffice us for now. I hope.

Reply:
I just found this PR and tried to read up. I have similar concerns as @bigwad - it would be really good if we could leave `Transaction.Current` out of this library, and also not dealing with RO-transactions in a special way. Let's see what we can do.

Reply:
As long as we can leave implicit read-only transactions to be in Palindrom only, [Replace TransactAndRebase extension with RebaseTransactor decorator #30](https://github.com/Starcounter/Starcounter.Palindrom/pull/30) pull request will be enough, and this one can be discarded.